File: tests/gpu/government_bonus_test.py
"""THE GOVERNMENT FLAT BONUS AND THE LEGACY CARD — the GPU half.

    npm run seed && npm run export        # (once) writes seeder/worlds/
    python tests/gpu/government_bonus_test.py

The TS twin is tests/cpu/seats/government-bonus.test.ts.

CIV6 (Gathering Storm): a government's second bonus is FLAT — eight
MODIFIER_PLAYER_GOVERNMENT_FLAT_BONUS rows and Communism's COMMUNISM_SCIENCE,
paid while the seat is IN the government. Expansion2_RemoveData.xml deletes
the base game's `*_ACCUMULATING` rows. The exporter writes a government's
inherent and flat bonus as ONE table row; a legacy card's row is its
government's inherent bonus alone.
"""

from __future__ import annotations

import sys
from pathlib import Path

import torch

sys.path.insert(0, str(Path(__file__).resolve().parent.parent.parent / "gpu"))
sys.path.insert(0, str(Path(__file__).resolve().parent))
from core import BatchSim, load_rules, load_fixture, fixture_paths
from warmup import settle_all

ROW = 0
SCIENCE = 3   # YIELD_KEYS: food, production, gold, science, culture, faith

# THE WARMED BASE: a scene pays a `restore` instead of a fixture load and a settle.
_BASE: dict = {}


def build(path) -> BatchSim:
    key = str(path)
    if key not in _BASE:
        sim = settle_all(BatchSim([load_fixture(path)], load_rules(),
                                  device="cpu", dtype=torch.float64))
        _BASE[key] = (sim, sim.snapshot())
    sim, snap = _BASE[key]
    sim.restore(snap)
    return sim


def gov_index(rules, gid: str) -> int:
    return [g["id"] for g in rules.governments].index(gid)


def in_government(sim: BatchSim, rules, civic: str) -> None:
    """the seat holds exactly one civic, the one that unlocks the government
    under test — `_adopted_gov` takes the newest tier it reaches"""
    import json
    from core import FIXTURES
    rj = json.loads((FIXTURES / "rules.json").read_text(encoding="utf-8"))
    ci = [c["id"] for c in rj["civics"]].index(civic)
    sim.civ_civics[:, ROW] = False
    sim.civ_civics[:, ROW, ci] = True
    sim._eff_version += 1


def test_the_wire(rules, path) -> None:
    """the flat bonus rides the government row; no legacy card row carries it"""
    sim = build(path)
    g = {gid: gov_index(rules, gid) for gid in ("AUTOCRACY", "FASCISM", "MONARCHY", "THEOCRACY",
                                                 "DEMOCRACY", "MERCHANT_REPUBLIC", "OLIGARCHY",
                                                 "CLASSICAL_REPUBLIC", "COMMUNISM", "CHIEFDOM")}
    assert float(sim._gov_prodb[g["AUTOCRACY"], 0]) == 1 and abs(float(sim._gov_prodb[g["AUTOCRACY"], 3]) - 0.1) < 1e-12, "Autocracy: +10% toward wonders"
    assert float(sim._gov_prodb[g["FASCISM"], 0]) == 2 and abs(float(sim._gov_prodb[g["FASCISM"], 3]) - 0.5) < 1e-12, "Fascism: +50% toward every unit"
    assert float(sim._gov_inflmult[g["MONARCHY"]]) == 1.5, "Monarchy: +50% influence"
    assert float(sim._gov_faithbuy[g["THEOCRACY"]]) == 15, "Theocracy: 15% off faith"
    assert float(sim._gov_goldbuy[g["DEMOCRACY"]]) == 15, "Democracy: 15% off gold"
    assert abs(float(sim._gov_distprod[g["MERCHANT_REPUBLIC"]]) - 1.15) < 1e-12, "Merchant Republic: +15% toward districts"
    assert float(sim._gov_xppct[g["OLIGARCHY"]]) == 20, "Oligarchy: +20% experience"
    assert abs(float(sim._gov_gppmult[g["CLASSICAL_REPUBLIC"]]) - 1.15) < 1e-12, "Classical Republic: +15% GPP"
    assert abs(float(sim._gov_ymult[g["COMMUNISM"], SCIENCE]) - 1.1) < 1e-12, "Communism: +10% science"
    leg = sim._pol_legacy >= 0
    assert int(leg.sum()) == sim._ngov - 1, "one legacy card per government but the Chiefdom"
    assert not bool((sim._pol_prodb[leg, 0] >= 0).any()), "a legacy card carries a flat production bonus"
    for t in (sim._pol_inflmult, sim._pol_distprod):
        assert bool((t[leg] == 1).all()), "a legacy card carries a flat multiplier"
    for t in (sim._pol_goldbuy, sim._pol_faithbuy):
        assert bool((t[leg] == 0).all()), "a legacy card carries a flat discount"
    print("  1 the wire OK — nine flat bonuses on the government rows, none on a legacy card")


def test_the_purchase_discounts(rules, path) -> None:
    """Democracy takes 15% off a GOLD purchase and nothing off faith;
    Theocracy the other way. `_gold_price` / `_faith_price` are `goldPrice` /
    `faithPrice`'s twins, applied where every purchase is priced and paid."""
    for civic, gid, purse, other, want in (("SUFFRAGE", "DEMOCRACY", "_gold_price", "_faith_price", 85.0),
                                          ("REFORMED_CHURCH", "THEOCRACY", "_faith_price", "_gold_price", 85.0)):
        sim = build(path)
        in_government(sim, rules, civic)
        adopted, _ = sim._adopted_gov(sim.civ_civics[:, ROW])
        assert int(adopted[0]) == gov_index(rules, gid), f"{civic} did not adopt {gid}"
        hundred = torch.full((sim.B,), 100.0, dtype=torch.float64)
        got = float(getattr(sim, purse)(ROW, hundred)[0])
        assert abs(got - want) < 1e-9, f"{gid}: 100 -> {got}, want {want}"
        assert float(getattr(sim, other)(ROW, hundred)[0]) == 100.0, f"{gid} touched the other purse"
        # a [1, N] price row broadcasts per game too: 50 less 15% is 42.5, floored to five
        row2 = getattr(sim, purse)(ROW, torch.tensor([[100.0, 50.0]], dtype=torch.float64))
        assert row2.shape == (sim.B, 2) and abs(float(row2[0, 1]) - 40.0) < 1e-9, row2
    print("  2 the purchase discounts OK — Democracy off gold, Theocracy off faith, 15% each")


def test_the_flat_bonus_is_the_governments_alone(rules, path) -> None:
    """Autocracy's +10% toward wonders is paid while the seat is IN Autocracy,
    and its legacy card slotted under another government pays the inherent
    bonus without it."""
    sim = build(path)
    in_government(sim, rules, "POLITICAL_PHILOSOPHY")
    prod = [p for p in sim._gov_mods(ROW)[12]["prod"] if bool(p[0][0])]
    assert any(p[1] == 1 and abs(p[4] - 0.1) < 1e-12 for p in prod), f"Autocracy's wonder bonus is not paid: {prod}"
    # every civic, every government held: a legacy card is slottable
    sim = build(path)
    sim.civ_civics[:, ROW] = True
    sim.civ_gov_held[:, ROW] = (1 << sim._ngov) - 1
    sim._eff_version += 1
    auto = gov_index(rules, "AUTOCRACY")
    p_idx = int((sim._pol_legacy == auto).nonzero()[0])
    before = sim._gov_mods(ROW)[12]["govbldy"].clone()
    chosen = torch.zeros(sim.B, sim._npol, dtype=torch.bool)
    chosen[:, p_idx] = True
    sim.seat_ext[:, ROW] = True
    sim.apply_seat_actions(ROW, policies=chosen)
    sim._seat_record_apply(ROW, torch.ones(sim.B, dtype=torch.bool))
    assert bool(sim._seat_slotted(ROW)[0, p_idx]), "the store did not take the legacy card"
    fx = sim._gov_mods(ROW)[12]
    assert float(fx["govbldy"][0] - before[0]) == float(sim._gov_govbldy[auto]), "the legacy card did not pay Autocracy's inherent bonus"
    assert not any(p[1] == 1 and abs(p[4] - 0.1) < 1e-12 for p in fx["prod"] if bool(p[0][0])), "the legacy card paid the flat bonus"
    print("  3 the flat bonus is the government's alone OK")


def main() -> int:
    rules = load_rules()
    path = fixture_paths()[0]
    test_the_wire(rules, path)
    test_the_purchase_discounts(rules, path)
    test_the_flat_bonus_is_the_governments_alone(rules, path)
    print("BATTERY OK government_bonus")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: tests/cpu/seats/government-bonus.test.ts
import { describe, it, expect } from 'vitest';
import { makeMap, makeState, settleAt, tileAtCoords, grantCivics } from '../helpers';
import { seatOf } from '../../../cpu/core/seats';
import { applySeatActionRecord } from '../../../cpu/core/phase';
import { computeAdoption, goldPrice, faithPrice, governmentBit, getModifiers, slottedPolicyIndices } from '../../../cpu/core/effects';
import { unitPurchaseCost } from '../../../cpu/core/game';
import { GOVERNMENTS, GOVERNMENT_LIST, POLICY_LIST } from '../../../cpu/data/policies';
import { CIVICS } from '../../../cpu/data/civics';
import type { GameState } from '../../../cpu/core/types';

/**
 * CIV6 (Gathering Storm): a government's SECOND bonus is flat — eight
 * MODIFIER_PLAYER_GOVERNMENT_FLAT_BONUS rows (Expansion1_Governments.xml,
 * re-shipped by Expansion2) and Communism's COMMUNISM_SCIENCE — paid while
 * the seat is IN the government. Gathering Storm deletes the base game's
 * accumulating rows (`*_ACCUMULATING`, Expansion2_RemoveData.xml) and
 * CAPABILITY_GOVERNMENTS_LEGACY_BONUSES with them. A legacy card pays the
 * government's INHERENT bonus (its `PolicyModifiers`), never the flat one.
 *
 * The purchase discounts land where a purchase is priced and paid
 * (`goldPrice` / `faithPrice`); the GPU twins are `_gold_price` /
 * `_faith_price`, pinned in tests/gpu/government_bonus_test.py.
 */
function scene(...civics: string[]): GameState {
  const state = makeState(makeMap(12, 12, 'GRASSLAND'));
  settleAt(state, tileAtCoords(state.map, 5, 5).index, 0);
  grantCivics(state, ...civics);
  return state;
}

describe('the flat government bonus', () => {
  it('is one row per government but the Chiefdom, and no legacy card carries it', () => {
    for (const g of GOVERNMENT_LIST) {
      if (g.tier === 0) {
        expect(g.bonus, 'the Chiefdom has no flat bonus').toBeUndefined();
        continue;
      }
      expect(g.bonus, `${g.id} has no flat bonus`).toBeDefined();
      const card = POLICY_LIST.find((p) => p.legacyOf === g.id)!;
      for (const k of Object.keys(g.bonus!)) expect(card.effects, `${g.id}'s legacy card pays ${k}`).not.toHaveProperty(k);
    }
    expect(GOVERNMENTS.AUTOCRACY.bonus).toEqual({ prodBoost: { target: 'wonder', classes: [], eraMax: -1, pct: 0.1 } });
    expect(GOVERNMENTS.FASCISM.bonus).toEqual({ prodBoost: { target: 'anyUnit', classes: [], eraMax: -1, pct: 0.5 } });
    expect(GOVERNMENTS.MERCHANT_REPUBLIC.bonus).toEqual({ districtProdMult: 1.15 });
    expect(GOVERNMENTS.COMMUNISM.bonus).toEqual({ yieldMult: { science: 1.1 } });
  });

  it("Democracy takes 15% off a gold purchase, and nothing off faith", () => {
    const state = scene('SUFFRAGE');
    expect(computeAdoption(seatOf(state, 0)!.research).government).toBe('DEMOCRACY');
    expect(goldPrice(state, 0, 100)).toBe(85);
    expect(faithPrice(state, 0, 100)).toBe(100);
    // the raw price composer is untouched — the discount lands at the purchase
    expect(unitPurchaseCost(state, 'WARRIOR', 0)).toBe(unitPurchaseCost(scene(), 'WARRIOR', 0));
  });

  it("Theocracy takes 15% off a faith purchase, and nothing off gold", () => {
    const state = scene('REFORMED_CHURCH');
    expect(computeAdoption(seatOf(state, 0)!.research).government).toBe('THEOCRACY');
    expect(faithPrice(state, 0, 200)).toBe(170);
    expect(goldPrice(state, 0, 200)).toBe(200);
  });

  it('is paid only while the seat is in the government', () => {
    const state = scene('POLITICAL_PHILOSOPHY');
    expect(computeAdoption(seatOf(state, 0)!.research).government).toBe('AUTOCRACY');
    expect(getModifiers(state, 0).prodBoosts).toContainEqual(GOVERNMENTS.AUTOCRACY.bonus!.prodBoost);
    const later = scene('SUFFRAGE');
    expect(getModifiers(later, 0).prodBoosts).not.toContainEqual(GOVERNMENTS.AUTOCRACY.bonus!.prodBoost);
  });
});

describe('what a legacy card pays', () => {
  it("its government's inherent bonus, and never the flat one", () => {
    const state = scene(...Object.keys(CIVICS));
    const s = seatOf(state, 0)!;
    s.government.held = GOVERNMENT_LIST.reduce((m, g) => m | governmentBit(g.id), 0);
    expect(computeAdoption(s.research).government).not.toBe('AUTOCRACY');
    const before = getModifiers(state, 0);
    const legIdx = POLICY_LIST.findIndex((p) => p.id === 'LEGACY_AUTOCRACY');
    applySeatActionRecord(state, s, { production: [], tech: null, civic: null, units: [], policies: [legIdx] });
    expect(slottedPolicyIndices(state, 0), 'the store did not take the legacy card').toContain(legIdx);
    const after = getModifiers(state, 0);
    expect(after.yieldsPerGovBuilding - before.yieldsPerGovBuilding).toBe(GOVERNMENTS.AUTOCRACY.effects.yieldsPerGovBuilding);
    expect(after.prodBoosts).not.toContainEqual(GOVERNMENTS.AUTOCRACY.bonus!.prodBoost);
  });
});
